src/trimerge-array.ts

```diff
@@ -0,0 +1,89 @@
+import { MergeFn } from './trimerge';
+import { Path } from './path';
+import { JSONObject, JSONValue } from './json';
+import { CannotMerge } from './cannot-merge';
+import { diff3Keys } from './diff3-keys';
+import { jsSameType } from './js-same-type';
+
+type ArrayKeyFn = (item: any, index: number, arrayPath: Path) => string;
+
+export function trimergeArrayCreator(
+  getArrayItemKey: ArrayKeyFn,
+  allowOrderConflicts: boolean = false,
+): MergeFn {
+  return (
+    orig: any[],
+    left: any[],
+    right: any[],
+    path: Path,
+    mergeFn: MergeFn,
+  ): JSONValue[] | typeof CannotMerge => {
+    if (jsSameType(orig, left, right) !== 'array') {
+      return CannotMerge;
+    }
+    const origMap: JSONObject = {};
+    const leftMap: JSONObject = {};
+    const rightMap: JSONObject = {};
+    const origKeys = orig.map((item, index): string => {
+      const key = getArrayItemKey(item, index, path);
+      if (key in origMap) {
+        throw new Error(`Duplicate array key '${key}' at /${path}`);
+      }
+      origMap[key] = item;
+      return key;
+    });
+    const leftKeys = left.map((item, index): string => {
+      const key = getArrayItemKey(item, index, path);
+      if (key in leftMap) {
+        throw new Error(`Duplicate array key '${key}' at /${path}`);
+      }
+      leftMap[key] = item;
+      return key;
+    });
+    const rightKeys = right.map((item, index): string => {
+      const key = getArrayItemKey(item, index, path);
+      if (key in rightMap) {
+        throw new Error(`Duplicate array key '${key}' at /${path}`);
+      }
+      rightMap[key] = item;
+      return key;
+    });
+
+    const mergedArray: JSONValue[] = [];
+    let leftSame = true;
+    let rightSame = true;
+    diff3Keys(
+      origKeys,
+      leftKeys,
+      rightKeys,
+      (key) => {
+        const leftElement = leftMap[key];
+        const rightElement = rightMap[key];
+        const merged = mergeFn(
+          origMap[key],
+          leftElement,
+          rightElement,
+          [...path, key],
+          mergeFn,
+        );
+        if (merged !== left[mergedArray.length]) {
+          leftSame = false;
+        }
+        if (merged !== right[mergedArray.length]) {
+          rightSame = false;
+        }
```

Comment:
Add a comment here explaining why these are comparing the merged value to the value at a specific index in `left` and `right`.

Reply:
Also, micro-optimization: could add `leftSame &&` and `rightSame &&` to the beginning of these conditions.

Reply:
I had that originally but wasn’t sure if it would be faster